File: src/test/scala/ru/tinkoff/gatling/kafka/examples/BasicSimulation.scala
package ru.tinkoff.gatling.kafka.examples

import io.gatling.core.Predef._
import io.gatling.core.feeder.Feeder
import io.gatling.core.session._
import io.gatling.core.structure.ScenarioBuilder
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.common.header.Header
import ru.tinkoff.gatling.kafka.Predef._
import ru.tinkoff.gatling.kafka.protocol.KafkaProtocol

import java.util.concurrent.atomic.AtomicInteger
import scala.concurrent.duration.DurationInt

class BasicSimulation extends Simulation {

  val kafkaConf: KafkaProtocol = kafka
    .topic("test.topic")
    .properties(Map(ProducerConfig.ACKS_CONFIG -> "1"))

  val kafkaProtocolC: KafkaProtocol = kafka.requestReply
    .producerSettings(
      Map(
        ProducerConfig.ACKS_CONFIG              -> "1",
        ProducerConfig.BOOTSTRAP_SERVERS_CONFIG -> "localhost:9092",
      ),
    )
    .consumeSettings(
      Map(
        "bootstrap.servers" -> "localhost:9092",
      ),
    )
    .timeout(5.seconds)
  val c                             = new AtomicInteger(0)
  val feeder: Feeder[Int]           = Iterator.continually(Map("kekey" -> c.incrementAndGet()))

  val scn: ScenarioBuilder = scenario("Basic")
    .feed(feeder)
    .exec(
      kafka("ReqRep").requestReply
        .requestTopic("test.t")
        .replyTopic("test.t")
        .send[String, String]("#{kekey}", """{ "m": "dkf" }""", List.empty[Header].expressionSuccess)
        .check(jsonPath("$.m").is("dkf")),
    )

  setUp(scn.inject(atOnceUsers(5))).protocols(kafkaProtocolC).maxDuration(120.seconds)

}

File: src/test/java/ru/tinkoff/gatling/javaapi/kafka/examples/BasicSimulation.java
package ru.tinkoff.gatling.javaapi.kafka.examples;

import io.gatling.javaapi.core.Simulation;
import org.apache.kafka.common.header.internals.RecordHeaders;
import ru.tinkoff.gatling.kafka.javaapi.protocol.*;
import org.apache.kafka.clients.producer.ProducerConfig;
import ru.tinkoff.gatling.kafka.javaapi.request.builder.RequestReplyBuilder;

import java.time.Duration;
import java.util.Map;

import static ru.tinkoff.gatling.kafka.javaapi.KafkaDsl.kafka;

public class BasicSimulation extends Simulation {

    // Имеется KafkaProtocolBuilderBase, он делится на не rr и requestReply
    // не requestReply
    public KafkaProtocolBuilder k1 = kafka().topic("").properties(Map.of(ProducerConfig.ACKS_CONFIG, "1"));
    // requestReply
    public KafkaProtocolBuilderNew r1 = kafka()
            .requestReply()
            .producerSettings(
                    Map.of(
                            ProducerConfig.ACKS_CONFIG, "1",
                            ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092"
                    )
            )
            .consumeSettings(
                    Map.of(
                            "bootstrap.servers", "localhost:9092"
                    )
            )
            .withDefaultTimeout()
            .matchByValue();

    public KafkaProtocolBuilderNew r2 = kafka()
            .requestReply()
            .producerSettings(
                    Map.of(
                            ProducerConfig.ACKS_CONFIG, "1",
                            ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092"
                    )
            )
            .consumeSettings(
                    Map.of(
                            "bootstrap.servers", "localhost:9092"
                    )
            )
            .timeout(Duration.ofSeconds(30));

    public RequestReplyBuilder<?, ?> krr1  = kafka("requestReply")
            .requestReply()
            .requestTopic("inputTopic")
            .replyTopic("outputTopic")
            .send("somekey", "someval", new RecordHeaders(), String.class, String.class);
}
